kernel/posix/pthread.c

```diff
@@ -0,0 +1,351 @@
+		pthread[thread_index].cancel_lock.sem =
+					&pthread[thread_index].cancel_lock_sem;
+		pthread_mutex_init(&pthread[thread_index].cancel_lock, &att);
+		pthread_cond_init(&pthread[thread_index].state_cond,
+				  &cond_attr);
+		*newthread = (pthread_t)
+				k_thread_create(&pthread[thread_index].thread,
+						attr->stack, attr->stacksize,
+						(k_thread_entry_t)
+						zephyr_thread_wrapper,
+						(void *)arg, NULL,
+						threadroutine, prio,
+						attr->options,
+						attr->delayedstart);
+
+	} else {
+		/* Insuffecient resource, maximum thread count reached */
+		errno = EAGAIN;
+		return -1;
+	}
+
+	return 0;
+}
+
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	unsigned int key;
+	struct posix_thread *pthread = (struct posix_thread *) pthread_self();
+
+	if (state != PTHREAD_CANCEL_ENABLE &&
+	    state != PTHREAD_CANCEL_DISABLE) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	*oldstate = pthread->cancel_state;
+	pthread->cancel_state = state;
+
+	key = irq_lock();
+
+	if (state == PTHREAD_CANCEL_DISABLE) {
+		_thread_essential_set();
+	} else {
+		_thread_essential_clear();
+	}
+
+	irq_unlock(key);
+
+	if (state == PTHREAD_CANCEL_ENABLE && pthread->cancel_pending) {
+		pthread_exit((void *)PTHREAD_CANCELED);
+	}
+
+	return 0;
+}
+
+int pthread_cancel(pthread_t pthread)
+{
+	struct posix_thread *thread = (struct posix_thread *) pthread;
+
+	if (thread == NULL || thread->state == PTHREAD_TERMINATED) {
+		errno = ESRCH;
+		return -1;
+	}
+
+	pthread_mutex_lock(&thread->cancel_lock);
+	thread->cancel_pending = 1;
+
+	if (thread->cancel_state == PTHREAD_CANCEL_ENABLE) {
+		if (thread->state == PTHREAD_DETACHED) {
+			thread->state = PTHREAD_TERMINATED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+			k_thread_abort((k_tid_t) thread);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+		}
+	}
+
+	return 0;
+}
+
+int pthread_setschedparam(pthread_t pthread, int policy,
+			  const struct sched_param *param)
+{
+	k_tid_t *thread = (k_tid_t *)pthread;
+	int new_prio, cur_prio;
+
+	if (thread == NULL) {
+		errno = ESRCH;
+		return -1;
+	}
+
+	if (policy != SCHED_RR && policy != SCHED_FIFO) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	new_prio = posix_to_zephyr_priority(param->priority, policy);
+	cur_prio = k_thread_priority_get(*thread);
+
+	if (is_posix_prio_valid(new_prio, policy) == false ||
+	    new_prio == cur_prio) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	k_thread_priority_set(*thread, new_prio);
+
+	return 0;
+}
+
+int pthread_attr_init(pthread_attr_t *attr)
+{
+	const pthread_attr_t init_pthread_attrs = {
+		.priority = K_LOWEST_APPLICATION_THREAD_PRIO,
+		.stack = NULL,
+		.stacksize = 0,
+		.options = PTHREAD_INIT_FLAGS,
+		.delayedstart = K_NO_WAIT,
+		.schedpolicy = SCHED_RR,
+		.detachstate = PTHREAD_CREATE_JOINABLE,
+	};
+
+	if (attr == NULL) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	memcpy((void *)attr, (const void *)&init_pthread_attrs,
```

Comment:
Are the casts necessary? `attr` is not const, and any pointer can be converted to `void`. `init_pthread_attrs` is const already, so it should be possible to pass it to the second memcpy() parameter. While the memcpy() is fine, you can write this as `*attr = init_pthread_attrs;`, and get rid of the casts.

Reply:
Yes I was using "*attr = init_pthread_attrs;" instead of memcpy earlier but I get feedback that memcpy will be better choice as compare to direct assignment.